// devextreme-angular/esm5/ui/nested/file-selection-item-dxi.js
/*!
 * devextreme-angular
 * Version: 19.2.6
 * Build date: Thu Jan 30 2020
 *
 * Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
/* tslint:disable:use-input-property-decorator */
import { Component, NgModule, Host, SkipSelf } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { DxiFileManagerToolbarItem } from './base/file-manager-toolbar-item-dxi';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'devextreme-angular/core';
var DxiFileSelectionItemComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxiFileSelectionItemComponent, _super);
    function DxiFileSelectionItemComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiFileSelectionItemComponent.prototype, "_optionPath", {
        get: function () {
            return 'fileSelectionItems';
        },
        enumerable: true,
        configurable: true
    });
    DxiFileSelectionItemComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxiFileSelectionItemComponent = tslib_1.__decorate([ tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxiFileSelectionItemComponent);
DxiFileSelectionItemComponent.ɵfac = function DxiFileSelectionItemComponent_Factory(t) { return new (t || DxiFileSelectionItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NestedOptionHost, 5), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NestedOptionHost, 1)); };
DxiFileSelectionItemComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DxiFileSelectionItemComponent, selectors: [["dxi-file-selection-item"]], inputs: { cssClass: "cssClass", disabled: "disabled", html: "html", locateInMenu: "locateInMenu", location: "location", menuItemTemplate: "menuItemTemplate", name: "name", options: "options", showText: "showText", template: "template", text: "text", visible: "visible", widget: "widget" }, features: [ɵngcc0.ɵɵProvidersFeature([NestedOptionHost]), ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxiFileSelectionItemComponent_Template(rf, ctx) { }, styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DxiFileSelectionItemComponent, [{
        type: Component,
        args: [{
                selector: 'dxi-file-selection-item',
                template: '',
                providers: [NestedOptionHost],
                inputs: [
                    'cssClass',
                    'disabled',
                    'html',
                    'locateInMenu',
                    'location',
                    'menuItemTemplate',
                    'name',
                    'options',
                    'showText',
                    'template',
                    'text',
                    'visible',
                    'widget'
                ],
                styles: ['']
            }]
    }], function () { return [{ type: ɵngcc1.NestedOptionHost, decorators: [{
                type: SkipSelf
            }, {
                type: Host
            }] }, { type: ɵngcc1.NestedOptionHost, decorators: [{
                type: Host
            }] }]; }, null); })();
    return DxiFileSelectionItemComponent;
}(DxiFileManagerToolbarItem));
export { DxiFileSelectionItemComponent };
var DxiFileSelectionItemModule = /** @class */ (function () {
    function DxiFileSelectionItemModule() {
    }
DxiFileSelectionItemModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DxiFileSelectionItemModule });
DxiFileSelectionItemModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DxiFileSelectionItemModule_Factory(t) { return new (t || DxiFileSelectionItemModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DxiFileSelectionItemModule, { declarations: [DxiFileSelectionItemComponent], exports: [DxiFileSelectionItemComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DxiFileSelectionItemModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DxiFileSelectionItemComponent
                ],
                exports: [
                    DxiFileSelectionItemComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return DxiFileSelectionItemModule;
}());
export { DxiFileSelectionItemModule };
